### DISCUSSION TITLE
MPDF CSS not working while converting to PDF

### DISCUSSION THREAD

**Discussion by cresttech** [mpdf/mpdf#2049] [General]
Hi 

I am creating  dynamic PDF using mpdf in CI4, When I am checking the HTML on browser, it's showing correct design, when converted same html to PDF, it's completely messed. Please see screenshot. 

https://prnt.sc/8xbIURxty4IB
https://prnt.sc/CqiBT9U02g7a

Here's my code

`public function generatePDF(){

        $flagPath = base_url() .'public/assets/images/';

        $defaultConfig = (new Mpdf\Config\ConfigVariables())->getDefaults();
        $fontDirs = $defaultConfig['fontDir'];

        $defaultFontConfig = (new Mpdf\Config\FontVariables())->getDefaults();
        $fontData = $defaultFontConfig['fontdata'];

        $mpdf_data = [
            'debug' => true,
            // 'allow_output_buffering' => true,
            // 'showImageErrors'  => true,
            // 'text_input_as_HTML' => true,
            
            'fontDir' => array_merge($fontDirs, [
                FCPATH . '/public/assets/fonts',
            ]),
            'fontdata' => $fontData + [ // lowercase letters only in font key
                'roboto' => [
                    'R' => 'Roboto-Regular.ttf',
                    'I' => 'Roboto-Regular.ttf',
                    'B' => 'Roboto-Bold.ttf',
                ]
            ],
            'default_font' => 'roboto'
        
        ];
        $mpdf = new \Mpdf\Mpdf($mpdf_data);
        
        $content = '
            <!DOCTYPE html>
            <html lang="en">
            <head>
            <title>Soccer You</title>
            <meta charset="utf-8">
            <meta name="viewport" content="width=device-width, initial-scale=1">
            <meta name="robots" content="noindex,nofollow,noarchive" />
            <link rel="preconnect" href="https://fonts.googleapis.com">
            <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
            <link href="https://fonts.googleapis.com/css2?family=Roboto:ital,wght@0,100;0,300;0,400;0,500;0,700;0,900;1,100;1,300;1,400;1,500;1,700;1,900&display=swap" rel="stylesheet">
            <style type="text/css">
            body { padding:0 !important; margin:0 !important; display:block !important; min-width:100% !important; width:100% !important; -webkit-text-size-adjust:none }
            * {box-sizing: border-box;}
            body {font-family: "roboto";font-size: 13px;}
            table {border-collapse: collapse;border:0;font-family: "roboto";}
            table td, table th {vertical-align: top;padding: 10px; text-align: left;}
            p {margin: 0;width: 100%;float: left; line-height: 1.5;}
            table.table_perform {border-collapse: separate; border-spacing: 2px;}
            table.table_perform tr td {background: #fff;}
            .table_perform th {font-size: 13px;padding: 10px 7px;}
            table.table_perform th img {width: 15px;vertical-align: middle;margin-right: 4px;float: left; margin-bottom: 10px;}
            table.table_transfer {border-collapse: separate; border-spacing: 2px;}
            table.table_transfer tr td {background: #fff;}
            table.table_transfer th img {width: 15px;vertical-align: middle;margin-right: 4px;float: left;}
            table.table_transfer td img{width: 20px;vertical-align: middle;margin-right: 5px;margin-top: -2px;}
            .transfer_history { width: 100%;background: #f8f8f8;padding: 12px;margin-top: 15px; }
            .performance_data { width: 100%;background: #f8f8f8;padding: 12px;margin-top: 15px; }
            .court_position .ground_fix{width: 100%;}
            .currnt_club {width: 100%;float: left;margin-top: 8px;}
            .currnt_club img {width: 35px;vertical-align: middle;margin-right: 5px;}
            .currnt_size {width: 100%;float: left;margin-top: 8px;}
            .currnt_size img {width: 25px;vertical-align: middle;margin-right: 5px;margin-top: -4px;}
            .h_width {width: 120px;float: left;}
            .bnr_img {width: 80%;margin-left: 0;margin-top: 15px; position: relative;z-index: -1;}
            .social_links {width: 160px;float: right;font-size: 13px;}
            .scl_hdr {width: 100%;float: left;margin-bottom: 8px;}
            .scl_hdr img {width: 22px;vertical-align: middle;margin-right: 5px;margin-top: -3px;}
            table.player_detail td.detail_list .dl_item {width: 100%;float: left;}
            table.player_detail td.detail_list .dl_item .dl_text {width: 100%;float: left;margin-top: 8px;}
            table.player_detail td.detail_list .dl_item .dl_text img {width: 24px;vertical-align: middle;margin-right: 4px;}
            table.player_detail td.detail_list {
                width: 33.33%;
                border-bottom: 1px solid #fff;
                padding: 10px 0;
            }
            table.player_detail {
                border-collapse: separate;
                border-spacing: 10px;
            }
            table.player_detail td.detail_list.no_border {
                border: 0;
            }
            .right_div {width: 100%; float: left;}
            .court_position {width: 100%;float: left;position: relative;margin-top: 10px;}
            .plyr_pos {width: 100%;float: left;}
            .plyr_pos .ppos_data {width: 130px;margin-top: 10px;}
            .dot_positions {position: absolute;left: 0;top: 0;width: 100%;height: 100%; }
            .dot_positions img { width: 17px;height: 17px;object-fit: contain;position: absolute; }
            img.pos_st {
                left: 50%;
                top: 5px;
                transform: translateX(-50%);
            }
            img.pos_cam {
                left: 50%;
                top: 34px;
                transform: translateX(-50%);
            }
            img.pos_cm {
                left: 50%;
                top: 53px;
                transform: translateX(-50%);
            }
            img.pos_lm {
                left: 33%;
                top: 29px;
            }
            img.pos_rm {
                right: 33%;
                top: 29px;	
            }
            img.pos_cdm {
                left: 50%;
                top: 73px;
                transform: translateX(-50%);
            }
            img.pos_lwb {
                left: 27%;
                top: 80px;
            }
            img.pos_lb {
                left: 34%;
                top: 101px;
            }
            img.pos_rb {
                right: 34%;
                top: 101px;
            }
            img.pos_rwb {
                right: 27%;
                top: 80px;
            }
            img.pos_gk {
                left: 50%;
                top: 133px;
                transform: translateX(-50%);
            }
            </style>
            </head>
            <body class="body" style="padding:0 !important; margin:0 !important; display:block !important; min-width:100% !important; width:100% !important;-webkit-text-size-adjust:none">
                <div width="100%" border="0" cellspacing="0" cellpadding="0">
                
                <div width="100%" border="0" cellspacing="0" cellpadding="0">
                    <div style="width:650px; min-width:650px; padding:0; margin:0 auto; font-weight:normal;">
                        <table width="100%" border="0" cellspacing="0" cellpadding="0" class="header_table">
                            <tr>
                                <td>
                                    <table width="100%" border="0" cellspacing="0" cellpadding="0" class="header">
                                        <tr>
                                            <td style="width: 45%;padding-left: 0;">
                                                <div style="width: 110%;float: left;margin-right: -10%;">
                                                    <img src="'.$flagPath.'logo.png" style="width: 120px;vertical-align: top;margin-top: -4px;">
                                                    <span style="font-size: 20px;padding-left: 5px;">PLAYER PROFILE</span>
                                                </div>
                                                <h1 style="width: 100%;float: left;font-size: 45px;margin: 20px 0;line-height: 1;">Elton Prince Morina</h1>
                                                <div style="width: 100%;float: left;">
                                                    <p style="font-size: 22px;">
                                                        <strong>Current Club</strong>
                                                        <span class="currnt_club" style="font-size: 20px;"><img src="'.$flagPath.'fc-logo.png"> FC Thun</span>
                                                    </p>
                                                </div>
                                                <div style="width: 100%;float: left;margin-top: 25px;">
                                                    <div class="h_width">
                                                        <strong>Height</strong>
                                                        <div class="currnt_size"><img src="'.$flagPath.'Icons19.png"> 175cm</div>
                                                    </div>
                                                    <div class="h_width">
                                                        <strong>Weight</strong>
                                                        <div class="currnt_size"><img src="'.$flagPath.'Icons18.png"> 72kg</div>
                                                    </div>
                                                </div>
                                            </td>
                                            <td style="width: 55%;padding-left: 0; padding-right: 0;">
                                                <div class="social_links">
                                                    <div class="scl_hdr"><img src="'.$flagPath.'Icons3.png">facebook.com/elton</div>
                                                    <div class="scl_hdr"><img src="'.$flagPath.'Icons2.png">instagram.com/elton</div>
                                                </div>
                                                <img src="'.$flagPath.'banner-pic.png" class="bnr_img">
                                            </td>
                                        </tr>
                                    </table>
                                </td>
                            </tr>
                            <tr>
                                <td style="background: #f8f8f8;padding: 0;">
                                    <table width="100%" border="0" cellspacing="0" cellpadding="0" class="player_detail">
                                        <tr>
                                            <td colspan="3" style="padding: 0;">
                                                <img src="'.$flagPath.'players-1.png" style="width: 24%;">
                                                <img src="'.$flagPath.'players-2.png" style="width: 24%;margin-left: 4px;">
                                                <img src="'.$flagPath.'players-3.png" style="width: 24%;margin-left: 4px;">
                                                <img src="'.$flagPath.'players-4.png" style="width: 24%;margin-left: 4px;">
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Age</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons20.png"> 20 Years</div>
                                                </div>
                                            </td>
                                            <td class="detail_list no_border">
                                                <div class="dl_item">
                                                    <strong>In the team since</strong>
                                                    <div class="dl_text">Apr 07, 2023</div>
                                                </div>
                                            </td>
                                            <td class="detail_list no_border">
                                                <div class="dl_item">
                                                    <strong>Top Speed</strong>
                                                    <div class="dl_text">33.7 km/h</div>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Nationality</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'flag-1.png"> Switzerland</div>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons8.png"> Kosovo</div>
                                                </div>
                                            </td>
                                            <td class="detail_list no_border">
                                                <div class="dl_item">
                                                    <strong>Current market value</strong>
                                                    <div class="dl_text">€54.5M</div>
                                                </div>
                                            </td>
                                            <td class="detail_list no_border">
                                                <div class="dl_item">
                                                    <strong>International Player</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons8.png"> Kosovo</div>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Date of birth</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons11.png"> Apr 21, 2024</div>
                                                </div>
                                            </td>
                                            <td colspan="2" class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Last change</strong>
                                                    <div class="dl_text">Apr 07, 2023</div>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Place of birth</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'flag-1.png"> Zurich, Switzerland</div>
                                                </div>
                                            </td>
                                            <td colspan="2" rowspan="4" class="detail_list no_border">
                                                <div class="court_position">
                                                    <img src="'.$flagPath.'Playground-project-c.png" class="ground_fix">
                                                    <div class="dot_positions">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_st">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_cam">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_lm">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_cm">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_rm">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_cdm">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_lwb">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_lb">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_rb">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_rwb">
                                                        <img src="'.$flagPath.'Playground-dot.png" class="pos_gk">
                                                    </div>
                                                </div>
                                                <div class="plyr_pos">
                                                    <p class="ppos_data"><strong>Main Position</strong><br />Forward</p>
                                                    <p class="ppos_data"><strong>Other position(s)</strong><br />Forward/Left/Right</p>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>Contract</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons15.png"> 2017 - 2025</div>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list">
                                                <div class="dl_item">
                                                    <strong>League</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons16.png"> Challenge League</div>
                                                </div>
                                            </td>
                                        </tr>
                                        <tr>
                                            <td class="detail_list no_border">
                                                <div class="dl_item">
                                                    <strong>Foot</strong>
                                                    <div class="dl_text"><img src="'.$flagPath.'Icons21.png"> Right</div>
                                                </div>
                                            </td>
                                        </tr>
                                    </table>
                                </td>
                            </tr>
                        </table>
                        <div class="transfer_history">
                            <h2 style="margin-top: 5px;padding: 0 10px;">Transfer History</h2>
                            <table width="100%" border="0" cellspacing="0" cellpadding="0" class="table_transfer">
                                <tr>
                                    <th><img src="'.$flagPath.'Icons6.png"><span>Saison</span></th>
                                    <th><img src="'.$flagPath.'Icons20.png"><span>Date</span></th>
                                    <th>Moving From</th>
                                    <th>Moving To</th>
                                </tr>
                                <tr style="background: #fff;">
                                    <td>2023</td>
                                    <td>19.07.2022</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> YF Juventus Zürich II</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> FC Thun, Switzerland</td>
                                </tr>
                                <tr style="background: #fff;">
                                    <td>2023</td>
                                    <td>19.07.2022</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> YF Juventus Zürich II</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> FC Thun, Switzerland</td>
                                </tr>
                                <tr style="background: #fff;">
                                    <td>2023</td>
                                    <td>19.07.2022</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> YF Juventus Zürich II</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> FC Thun, Switzerland</td>
                                </tr>
                                <tr style="background: #fff;">
                                    <td>2023</td>
                                    <td>19.07.2022</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> YF Juventus Zürich II</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> FC Thun, Switzerland</td>
                                </tr>
                                <tr style="background: #fff;">
                                    <td>2023</td>
                                    <td>19.07.2022</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> YF Juventus Zürich II</td>
                                    <td style="white-space: nowrap;"><img src="'.$flagPath.'fcsicon-icon.png"> FC Thun, Switzerland</td>
                                </tr>
                            </table>
                        </div>
                        <div class="performance_data">
                            <h2 style="margin-top: 5px;padding: 0 10px;">Performance Data</h2>
                            <table width="100%" border="0" cellspacing="0" cellpadding="0" class="table_perform">
                                <tr>
                                    <th><img src="'.$flagPath.'flag-1.png" style="margin-top: 3px;"><span>Team</span></th>
                                    <th><img src="'.$flagPath.'Icons6.png"><span>Saison</span></th>
                                    <th><img src="'.$flagPath.'Icons13.png"><span>Matches</span></th>
                                    <th><img src="'.$flagPath.'Icons17.png"><span>Goals</span></th>
                                    <th><img src="'.$flagPath.'Icons14.png"><span>Coach<br /> during debut</span></th>
                                    <th><img src="'.$flagPath.'Icons7.png"><span>Age of<br /> player</span></th>
                                </tr>
                                <tr>
                                    <td>FC Thun u 21</td>
                                    <td>23/24</td>
                                    <td>5</td>
                                    <td>3</td>
                                    <td>Stipe Matic</td>
                                    <td>19 Years</td>
                                </tr>
                                <tr>
                                    <td>YF Juventus Zürich II</td>
                                    <td>23/24</td>
                                    <td>5</td>
                                    <td>3</td>
                                    <td>Piero Bauer</td>
                                    <td>18 Years</td>
                                </tr>
                                <tr>
                                    <td>GC U18</td>
                                    <td>23/24</td>
                                    <td>5</td>
                                    <td>3</td>
                                    <td>Injured 8 Month</td>
                                    <td>18 Years</td>
                                </tr>
                                <tr>
                                    <td>GC U18</td>
                                    <td>23/24</td>
                                    <td>5</td>
                                    <td>3</td>
                                    <td>Heris Stefanachi</td>
                                    <td>18 Years</td>
                                </tr>
                                <tr>
                                    <td>FC Rapperswil-Jona</td>
                                    <td>23/24</td>
                                    <td>5</td>
                                    <td>3</td>
                                    <td>Urs Wolfensberger</td>
                                    <td>17 Years</td>
                                </tr>
                            </table>
                        </div>
                        <table width="100%" border="0" cellspacing="0" cellpadding="0" class="footer_table">
                            <tr>
                                <td style="text-align: center;font-weight: bold;">Succer You Sports AG | www.socceryou.ch </td>
                            </tr>
                        </table>
                    </div>
                </div>
            </div>
            </body>
            </html>
        ';

        $stylesheet = file_get_contents(FCPATH .'public/assets/css/user-profile-pdf.css'); // external css

        // $mpdf->WriteHTML($content);

        $mpdf->WriteHTML($stylesheet,\Mpdf\HTMLParserMode::HEADER_CSS);
        $mpdf->WriteHTML($content,\Mpdf\HTMLParserMode::HTML_BODY);

        $this->response->setHeader('Content-Type', 'application/pdf');

        // $mpdf->Output( WRITEPATH . 'uploads/documents/soccer-pdf.pdf','F'); // opens in browser
        // $mpdf->Output( WRITEPATH . 'uploads/documents/soccer-pdf.pdf','I'); // opens in browser

        $mpdf->Output(); 
        // exit;

    }`

Please let me know how can I fix this?
thanks

**Comment by ssieb** [comment]
Have you checked that the CSS you're using is supported?

**Reply by cresttech** [reply]
Yes, I changed text color, and color was changed in PDF, so CSS is adding in PDF
https://prnt.sc/CaMVQfkvRFSh

**Reply by ssieb** [reply]
That's not what I mean.  The CSS will be processed, but you are probably using declarations that aren't supported by MPDF.  See the list of what is and isn't supported.

**Reply by cresttech** [reply]
did you mean these limitations?
https://mpdf.github.io/about-mpdf/limitations.html

**Reply by ssieb** [reply]
More this: https://mpdf.github.io/css-stylesheets/supported-css.html

**Reply by cresttech** [reply]
okay thanks... let me cross my HTML.

**Reply by manish2120** [reply]
@cresttech is it now working??
